notebook/notebookapp.py

```diff
@@ -674,11 +676,16 @@ def _default_cookie_secret_file(self):
     def _default_cookie_secret(self):
         if os.path.exists(self.cookie_secret_file):
             with io.open(self.cookie_secret_file, 'rb') as f:
-                return f.read()
+                key =  f.read()
         else:
-            secret = encodebytes(os.urandom(1024))
-            self._write_cookie_secret_file(secret)
-            return secret
+            key = encodebytes(os.urandom(1024))
+            self._write_cookie_secret_file(key)
+        h = sha256(digest_size=len(key))
+        h.update(key)
+        h.update(self.password.encode())
```

Comment:
IMO this should be a [HMAC](https://docs.python.org/3/library/hmac.html), with the secret key as the HMAC key & the password as the message. HMACs are built for this exact specific use case, and supported in the stdlib.

Reply:
I disagree that HMAC are built for that (no code is going to check hmac(key, pw) equal the value here). The result of this function will be used by HMAC as the key later on (in tornado), so we're basically doing `HMAC(HMAC(key, pwd_sha), cookie_value)`; Though I agree that the returned value with HMAC will still be (pseudo-)random enough.